lib/token.js

```diff
@@ -0,0 +1,276 @@
+'use strict';
+
+const types = require('./types');
+const util = require('util');
+
+const _Murmur3TokenType = types.dataTypes.getByName('bigint');
+const _RandomTokenType = types.dataTypes.getByName('varint');
+const _OrderedTokenType = types.dataTypes.getByName('blob');
+
+/**
+ * Represents a token on the Cassandra ring.
+ *
+ * @property value The raw value of the token.  Type depends on
+ *                 partitioner implementation. 
+ * 
+ */
+class Token {
+  constructor(value, tokenizer) {
+    this.value = value;
+    Object.defineProperty(this, '_tokenizer', { value: tokenizer, enumerable: false});
+  }
+
+  /**
+   * @returns {{code: number, info: *|Object}} The type info for the
+   *                                           type of the value of the token.
+   */
+  getType() {
+    throw new Error('You must implement a getType function for this Token instance');
+  }
+
+  /**
+   * @returns {*} The raw value of the token.
+   */
+  getValue() {
+    return this.value;
+  }
+
+  /**
+   * Encodes the underlying value into a buffer that can be
+   * used for routing.
+   * 
+   * @param {Encoder} encoder Encode for encoding to buffer.
+   * @returns {Buffer} buffer data for the token.
+   */
+  toBuffer(encoder) {
+    return encoder.encode(this.getValue(), this.getType());
+  }
+
+  toString() {
+    return this._tokenizer.stringify(this.getValue());
+  }
+
+  /**
+   * Returns 0 if the values are equal, 1 if greater than other, -1
+   * otherwise.
+   *
+   * @param {Token} other 
+   * @returns {Number}
+   */
+  compare(other) {
+    return this._tokenizer.compare(this.getValue(), other.getValue());
+  }
+
+  equals(other) {
+    return this.compare(other) === 0;
+  }
+}
+
+/**
+ * Represents a token from a Cassandra ring where the partitioner
+ * is Murmur3Partitioner.
+ * 
+ * The raw token type is a varint (represented by MutableLong).
+ */
+class Murmur3Token extends Token {
```

Comment:
Should `Murmur3Token` / `RandomToken` / `ByteOrderedToken` be exported to public?  The java driver doesn't do it, but I also don't see the harm in it and might be nice for instanceof checks (although probably not likely to be used).

Reply:
I'd say we only export `Token` but not the actual implementations.